hail/python/hailtop/utils/utils.py

```diff
@@ -407,11 +448,20 @@ async def __aexit__(self,
                         exc_tb: Optional[TracebackType]) -> None:
         self._subsema.release()
 
-        _, exc, _ = sys.exc_info()
-        if exc:
-            await self._shutdown()
-        else:
-            await self.wait_done()
+        if exc_val:
+            if self._exception is None:
+                self._exception = exc_val
+                await self._shutdown()
+            else:
+                log.info('discarding exception', exc_info=exc_val)
+
+        await self._done_event.wait()
```

Comment:
Is this first wait supposed to be here? Otherwise, all the code is the same except for the refactoring.

Reply:
OK, I changed this a bit to be clearer and I'm happier with it.  The invariant is:

_done_event is set if (1) we are not shut down and _pending is empty, or (2) we did shutdown and all the cancelled tasks are complete.  That means we can be shutdown, _pending can be empty, but _done_event is not (yet) set.

It also means _done_event can be set and then cleared again, if the last task completes but another one is then submitted.

Then I changed the code to maintain this invariant.  If we add a task to _pending, I clear _done_event.  If the last pending task finishes, set it.  In shutdown, if all the cancelled tasks are complete, set it, etc.

Finally, __aexit__ needs to wait for _event_done to be set _and_ pending to be empty (or None).

Make sense?